File: VoiceDropApp/RecordingName.swift
import Foundation

/// Builds the enriched, ASCII-only recording filename. Keeps the `VoiceDrop-`
/// prefix and `.m4a` suffix (the mining pipeline filters on those), and packs
/// readable context between them so a file is self-describing in a listing:
///
///   VoiceDrop-2026-06-18-143052-0m33s-Thu-Afternoon-Shanghai-Xuhui.m4a
///
/// Everything is ASCII (letters, digits, hyphens) — no spaces, no CJK, no
/// punctuation — so it round-trips cleanly through URLs, R2 keys and curl.
enum RecordingName {

    /// `place` is an already-clean ASCII tag like "Shanghai-Xuhui" (or nil).
    static func make(start: Date, duration: TimeInterval, place: String?) -> String {
        var parts = ["VoiceDrop", timestamp(start), durationTag(duration), weekday(start), period(start)]
        if let place, !place.isEmpty { parts.append(place) }
        return parts.joined(separator: "-") + ".m4a"
    }

    /// The decoded fields of a recording filename — the inverse of `make`. This is the
    /// SINGLE place that knows the token layout; every reader pulls fields off `Parsed`
    /// instead of re-splitting on "-" and hard-coding positional indices (which silently
    /// break the moment `make` adds/reorders a field).
    struct Parsed {
        let sessionTs: String   // "yyyy-MM-dd-HHmmss" (== make's timestamp)
        let month: Int?
        let day: Int?
        let hhmm: String?       // "HH:mm"
        let duration: String?   // "0m33s"
        let place: String?      // district if present, else city
    }

    /// Parse a `VoiceDrop-…` stem (filename without `.m4a`). Returns nil if it doesn't
    /// match the convention. Mirrors `make`: parts = VoiceDrop, yyyy, MM, dd, HHmmss,
    /// dur, weekday, period, [city, [district]].
    static func parse(_ stem: String) -> Parsed? {
        let p = stem.components(separatedBy: "-")
        guard p.count >= 5, p[0] == "VoiceDrop", p[1].count == 4 else { return nil }
        let sessionTs = p[1...4].joined(separator: "-")
        var hhmm: String?
        if p[4].count == 6 { hhmm = "\(p[4].prefix(2)):\(p[4].dropFirst(2).prefix(2))" }
        let duration = p.first { $0.range(of: #"^\d+m\d+s$"#, options: .regularExpression) != nil }
        let place = p.count >= 10 ? p[9] : (p.count >= 9 ? p[8] : nil)
        return Parsed(sessionTs: sessionTs, month: Int(p[2]), day: Int(p[3]), hhmm: hhmm, duration: duration, place: place)
    }

    /// True if `name` is a finished recording file (`VoiceDrop-….m4a`). The ONE
    /// predicate — was inlined (and slightly divergent) in LibraryStore + Uploader.
    static func isRecordingFile(_ name: String) -> Bool {
        name.hasPrefix("VoiceDrop-") && name.hasSuffix(".m4a")
    }

    static func timestamp(_ d: Date) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd-HHmmss"
        return f.string(from: d)
    }

    /// Parse a `yyyy-MM-dd-HHmmss` stamp (a sessionTs) back to a Date — used to
    /// compute a photo's offset (seconds since the recording started).
    static func date(fromTimestamp ts: String) -> Date? {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd-HHmmss"
        return f.date(from: ts)
    }

    /// The R2 relative key for a scene photo: `photos/<sessionTs>/<offset>-<rand>.jpg`.
    /// - `sessionTs` groups every photo of one recording (correlation + cheap cleanup).
    /// - `offset` = seconds from the recording start to capture — this IS "第几秒加进来"
    ///   (absolute capture time is recoverable as sessionTs + offset, so nothing is lost),
    ///   and it's far shorter than the old absolute `yyyy-MM-dd-HHmmss` capture stamp.
    /// - a 3-char base36 tail makes the key unique even when several photos land in the
    ///   SAME offset-second (e.g. a 9-photo album import, whose callbacks fire near-
    ///   simultaneously) — the old seconds-only name silently overwrote and lost photos.
    static func photoKey(sessionTs: String, offset: Int) -> String {
        "photos/\(sessionTs)/\(max(0, offset))-\(randomTag()).jpg"
    }

    /// `count` random base36 chars (0-9a-z). 3 → 46,656 combos: even 9 photos sharing
    /// one offset-second collide with ~0.08% probability — comfortably safe, still tiny.
    static func randomTag(_ count: Int = 3) -> String {
        let alphabet = Array("0123456789abcdefghijklmnopqrstuvwxyz")
        return String((0..<count).map { _ in alphabet.randomElement()! })
    }

    static func durationTag(_ t: TimeInterval) -> String {
        let total = max(0, Int(t.rounded()))
        return "\(total / 60)m\(total % 60)s"
    }

    private static let weekdays = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]

    static func weekday(_ d: Date) -> String {
        let wd = Calendar.current.component(.weekday, from: d)   // 1 = Sunday
        return weekdays[(wd - 1) % 7]
    }

    static func period(_ d: Date) -> String {
        switch Calendar.current.component(.hour, from: d) {
        case 5..<9:   return "EarlyMorning"
        case 9..<12:  return "Morning"
        case 12..<14: return "Noon"
        case 14..<18: return "Afternoon"
        case 18..<20: return "Evening"
        case 20..<23: return "Night"
        default:      return "LateNight"
        }
    }
}

File: android/app/src/main/kotlin/com/wangjianshuo/voicedrop/RecordingName.kt
package com.wangjianshuo.voicedrop

import java.text.SimpleDateFormat
import java.util.*

object RecordingName {
    private val dateFormatter = SimpleDateFormat("yyyyMMddHHmmss", Locale.US)
    private val weekdayNames = listOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")
    private val periodNames = listOf("am", "am", "am", "pm", "pm", "pm")

    fun stagingName(): String {
        val now = dateFormatter.format(Date())
        return "recording-$now.m4a"
    }

    fun promotedName(durationSec: Int, city: String? = null, district: String? = null): String {
        val now = Date()
        val ts = dateFormatter.format(now)
        val cal = Calendar.getInstance()
        cal.time = now
        val weekday = weekdayNames.getOrElse(cal.get(Calendar.DAY_OF_WEEK) - 1) { "周一" }
        val hour = cal.get(Calendar.HOUR_OF_DAY)
        val period = when (hour) {
            in 0..4 -> periodNames[0]
            in 5..7 -> periodNames[1]
            in 8..10 -> periodNames[2]
            in 11..12 -> periodNames[3]
            in 13..17 -> periodNames[4]
            else -> periodNames[5]
        }
        val location = when {
            city != null && district != null -> "-$city-$district"
            city != null -> "-$city"
            else -> ""
        }
        return "VoiceDrop-$ts-${durationSec}-$weekday-$period$location.m4a"
    }

    fun stem(audioName: String): String {
        return audioName.substringBeforeLast(".m4a").substringBeforeLast(".empty")
    }

    fun sessionTs(audioName: String): String? {
        val parts = audioName.removeSuffix(".m4a").split("-")
        return if (parts.size >= 5) {
            parts.slice(1..4).joinToString("-")
        } else null
    }

    fun date(fromTimestamp: String): Date? {
        return try {
            dateFormatter.parse(fromTimestamp)
        } catch (_: Exception) { null }
    }

    fun photoKey(sessionTs: String, offset: Int): String {
        val rand = (1..46656).random().toString(36)
        return "photos/$sessionTs/$offset-$rand.jpg"
    }
}
